### PR TITLE
inject literal config values

### PR DESCRIPTION
**Pull request by turbocrime**
use webpack config to inject literals and environment variables

for example with `MINIFRONT_URL=https://localhost:5173 pnpm dev`

the typescript

```ts
console.log(
  'displaying injected literals',

  process.env['NODE_ENV'],

  PRAX,
  PRAX_ORIGIN,

  IDB_VERSION,
  USDC_ASSET_ID,

  MINIFRONT_URL,
  DEFAULT_GRPC_URL,
);

```

may become the javascript

```js
console.log(
  'displaying injected literals',
  'development',
  'lkpmkhpnhknhmibgnmmhdhgdilepfghe',
  'chrome-extension://lkpmkhpnhknhmibgnmmhdhgdilepfghe',
  23,
  'reum7wQmk/owgvGMWMZn/6RFPV24zIKq3W6In/WwZgg=',
  'https://localhost:5173',
  'https://grpc.testnet.penumbra.zone/',
);
```

### CODE REVIEW

File: apps/extension/webpack.config.js
Comment (grod220) [inline apps/extension/webpack.config.js:1]: Can we convert this to a typescript file? `webpack.config.ts`?

File: packages/constants/config.ts
Comment (grod220) [inline packages/constants/config.ts:1]: I think it's worthwhile keeping a `prod` an `testnet` config that the web app webpack.config.js can import. It's helpful to keep our hardcoded registries and parameter stuff in one place.

File: apps/extension/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      (argv.mode == 'development' && process.env.MINIFRONT_URL) || MINIFRONT_URL,
+    ),
+
+    DEFAULT_GRPC_URL: JSON.stringify(
+      // you may want https://grpc.testnet-preview.penumbra.zone/
+      (argv.mode == 'development' && process.env.PENUMBRA_NODE_PD_URL) || PENUMBRA_NODE_PD_URL,
+    ),
+  };
+
+  console.log('webpack', { argv, definitions });
```
Comment (grod220) [inline apps/extension/webpack.config.js:38]: can remove

File: apps/extension/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+export default (env, argv) => {
+  // types declared in prax.d.ts
+  const definitions = {
+    // process.env.NODE_ENV is automatically provided by DefinePlugin
+
+    PRAX: JSON.stringify(PRAX),
+    PRAX_ORIGIN: JSON.stringify(`chrome-extension://${PRAX}`),
+
+    IDB_VERSION: JSON.stringify(IDB_VERSION),
```
Comment (grod220) [inline apps/extension/webpack.config.js:25]: Is this a type mismatch if `IDB_VERSION` is expect to be a number?

File: apps/extension/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+const __dirname = new URL('.', import.meta.url).pathname;
+const srcDir = path.join(__dirname, 'src');
+
+export default (env, argv) => {
+  // types declared in prax.d.ts
+  const definitions = {
+    // process.env.NODE_ENV is automatically provided by DefinePlugin
+
+    PRAX: JSON.stringify(PRAX),
+    PRAX_ORIGIN: JSON.stringify(`chrome-extension://${PRAX}`),
```
Comment (grod220) [inline apps/extension/webpack.config.js:23]: the stringify's feel unnecessary if they are supposed to be strings already

File: apps/extension/webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+import HtmlWebpackPlugin from 'html-webpack-plugin';
+import webpack from 'webpack';
+
+const PRAX = 'lkpmkhpnhknhmibgnmmhdhgdilepfghe';
+
+const IDB_VERSION = 23;
+const USDC_ASSET_ID = 'reum7wQmk/owgvGMWMZn/6RFPV24zIKq3W6In/WwZgg=';
+
+const MINIFRONT_URL = 'https://app.testnet.penumbra.zone/';
+const PENUMBRA_NODE_PD_URL = 'https://grpc.testnet.penumbra.zone/';
```
Comment (grod220) [inline apps/extension/webpack.config.js:12]: See my other comment, think we should import these

File: apps/extension/webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        // Required by the `bip39` library
+        Buffer: ['buffer', 'Buffer'],
+      }),
+      new webpack.IgnorePlugin({
+        // Not required by the `bip39` library, but very nice
+        checkResource(resource) {
+          return /.*\/wordlists\/(?!english).*\.json/.test(resource);
+        },
+      }),
+      new webpack.DefinePlugin(definitions),
```
Comment (grod220) [inline apps/extension/webpack.config.js:116]: We could just do the prod/dev diff here:

```typescript
new webpack.DefinePlugin(argv.mode == 'prod' ? prodDefinitions : devDefinitions)
```

**Review by grod220 — COMMENTED** [review]
Appears to address https://github.com/penumbra-zone/web/issues/139 👍

File: apps/extension/package.json [hunk 6]
```diff
@@ -5,9 +5,9 @@
   "type": "module",
   "scripts": {
     "download-keys": "tsx src/utils/download-proving-keys.ts",
-    "dev": "webpack --config webpack/webpack.dev.js --watch",
+    "dev": "webpack --watch --mode=development -d inline-source-map",
     "clean": "rm -rfv dist bin",
-    "build": "webpack --config webpack/webpack.prod.js",
+    "build": "webpack --mode=development -d inline-source-map",
```
Comment (grod220) [inline apps/extension/package.json:10]: build should be `production` no?

File: apps/extension/webpack.config.js
Comment (turbocrime) [reply apps/extension/webpack.config.js:1]: we could, but it's really annoying, and i don't want to do it in this pr

File: apps/extension/webpack.config.js (hunk 1 above)
Comment (turbocrime) [reply apps/extension/webpack.config.js:38]: this echoes webpack parameters including environment variables that are affecting config to the terminal, which i think is good.

File: apps/extension/webpack.config.js (hunk 2 above)
Comment (turbocrime) [reply apps/extension/webpack.config.js:25]: it's injected as a literal string `23` which is a number in javascript

File: apps/extension/webpack.config.js (hunk 3 above)
Comment (turbocrime) [reply apps/extension/webpack.config.js:23]: defineplugin needs a string to inject https://webpack.js.org/plugins/define-plugin/

File: apps/extension/package.json (hunk 6 above)
Comment (turbocrime) [reply apps/extension/package.json:10]: this is a script for developer convenience. if a develper is running `pnpm build` at the terminal, they are probably doing development, and want development features. in the unsual case they might want something else, the developer can run webpack directly and more explicitly.

in this PR, when CI makes a production build, it uses 

https://github.com/penumbra-zone/web/blob/43bf771a567a3705554fc13cd1ec9f038e9589e3/.github/workflows/extension-publish.yml#L41-L42

File: apps/extension/webpack.config.js (hunk 4 above)
Comment (turbocrime) [reply apps/extension/webpack.config.js]: now using dotenv to import

File: apps/extension/webpack.config.js (hunk 5 above)
Comment (turbocrime) [reply apps/extension/webpack.config.js:108]: removed any prod/dev switching

File: apps/extension/package.json (hunk 6 above)
Comment (turbocrime) [reply apps/extension/package.json]: okay, i changed this to just run `webpack` with no options, which defaults to `--mode=production` with no `devtool` (equivalent to before). the change to `extension-publish.yml` was reverted

File: apps/extension/webpack.config.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+// environment variables, you will override those defaults.
+dotenv.config();
+
+const definitions = {
+  // process.env.NODE_ENV is automatically provided by DefinePlugin
+
+  PRAX: JSON.stringify(process.env.PRAX),
+  PRAX_ORIGIN: JSON.stringify(`chrome-extension://${process.env.PRAX}`),
+
+  IDB_VERSION: JSON.stringify(Number(process.env.IDB_VERSION)),
```
Comment (grod220) [inline apps/extension/webpack.config.js:17]: Why would we need to go from string -> number -> string?

**Review by grod220 — APPROVED** [review]
Just a few questions, but feel free to merge after

File: apps/extension/webpack.config.js (hunk 7 above)
Comment (turbocrime) [reply apps/extension/webpack.config.js:17]: we could do it directly, but this provides a little bit of type safety, and prevents any creative escape sequence adventures

File: apps/extension/webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+// Loads default vars from `.env` file in this directory.  If you set
+// environment variables, you will override those defaults.
+dotenv.config();
+
+const definitions = {
+  // process.env.NODE_ENV is automatically provided by DefinePlugin
+
+  PRAX: JSON.stringify(process.env.PRAX),
+  PRAX_ORIGIN: JSON.stringify(`chrome-extension://${process.env.PRAX}`),
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply apps/extension/webpack.config.js:15]: they are not equal - again please see defineplugin docs https://webpack.js.org/plugins/define-plugin/#usage

[the remark this answers is not in the document]
Comment (turbocrime) [reply apps/extension/webpack.config.js:15]: tldr it injects a *code string* - so directly defining a string as 

```
SOME_DEFINE: `chrome-extension://${process.env.PRAX}`
```

would transform the code

```
console.log("Some define:", SOME_DEFINE);
```

into 

```
console.log("Some define:", chrome-extension://whatever);
```

when what you want is 


```
console.log("Some define:", "chrome-extension://whatever");
```

**Pull request merged**